### PR TITLE
[lsp] added a server socket launcher

### PR DESCRIPTION
**Pull request by JanKoehnlein**
For debugging, it is often useful to start LSs behind a socket server. This class allows for it.

### CODE REVIEW

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/SocketServerLauncher.xtend [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+ * 
+ * Useful for debugging live language servers.
+ * 
+ * @author Jan K&ouml;hnlein - Initial contribution and API
+ * @since 2.18
+ */
+@Log
+class SocketServerLauncher {
+
+	static val HOST = '-host'
```
Comment (cdietrich) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/SocketServerLauncher.xtend]: maybe should be made public analogous to ServerLauncher

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/SocketServerLauncher.xtend [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	static val PORT = '-port'
+	static val TRACE = '-trace'
+	static val NO_VALIDATE = '-noValidate'
```
Comment (cdietrich) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/SocketServerLauncher.xtend]: should there be a log option analogous to server laucnher

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/SocketServerLauncher.xtend [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+				launcher.startListening
+				LOG.info("Xtext Language Server has been started.")
+			}
+		} catch (Throwable throwable) {
+			throwable.printStackTrace()
+		}
+	}
+
+	protected def getServerModule() {
+		new ServerModule
```
Comment (cdietrich) [inline org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/SocketServerLauncher.xtend]: i like that. unfortunately we dont have it in normal server launcher

Comment (cdietrich) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/SocketServerLauncher.xtend]: but does it actually help (we still call the consturcutor of the subclass, so the name need to be duplicated again

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/SocketServerLauncher.xtend (hunk 2 above)
Comment (JanKoehnlein) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/SocketServerLauncher.xtend]: Logging is always enabled, as this is a debug scenario. In the server launcher the `log` option mainly serves for redirecting the logs to not pollute`stdin` and `stdout` which are used for JSON-RPC. I don't think these have to be 100% symmetric.

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/SocketServerLauncher.xtend (hunk 1 above)
Comment (JanKoehnlein) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/SocketServerLauncher.xtend]: I cannot think of any usecase where I really want to access that from the outside.

Comment (cdietrich) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/SocketServerLauncher.xtend]: i assume the usecase is subclassing (am not sure why they are public in server launcher)

File: org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/SocketServerLauncher.xtend (hunk 3 above)
Comment (JanKoehnlein) [reply org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/SocketServerLauncher.xtend]: We could improve on the `ServerLauncher` as well. 
Ah, I missed to add a return type!

**Comment by JanKoehnlein** [comment]
Am I right that CI errors are not related to this change?

**Comment by cdietrich** [comment]
which ci errors do you relate to?

**Comment by JanKoehnlein** [comment]
@cdietrich https://ci-staging.eclipse.org/xtext/job/xtext-core/job/jk%252Fsocketserver/

**Comment by cdietrich** [comment]
cc @kthoms @Hejado  ????
also we get timeouts at ci.eclipse.org/xtext
which are new to me.

**Comment by kthoms** [comment]
Please refer to https://ci.eclipse.org/xtext, not to *ci-staging*. We are currently making builds ready for JIRO. There are chances that the jobs there are not always working yet.

**Comment by kthoms** [comment]
https://ci.eclipse.org/xtext/job/xtext-core/job/jk%252Fsocketserver/

**Comment by JanKoehnlein** [comment]
If it's not CI, it should better not prevent merges on Github.

**Comment by cdietrich** [comment]
@JanKoehnlein i would have preferred to see at least a single green build

**Comment by szarnekow** [comment]
There were two timeouts on the branch build and now there is a timeout on master.

**Comment by cdietrich** [comment]
i will trigger the foundation

**Comment by kthoms** [comment]
@JanKoehnlein Please do not merge before approval! Why the hurry???

**Pull request merged**